Refactor the authorization task (#3150)
istio/istio.io: content/docs/tasks/security/authz-permissive/index.md

you to verify authorization policies before applying them in a production environment.

The authorization permissive mode is an experimental feature in version 1.1. Its interface can change
in future releases. If you do not want to try out the permissive mode feature, you can directly
[enable Istio authorization](/docs/tasks/security/authz-http#enabling-istio-authorization) to skip
enabling the permissive mode.

This task covers two scenarios regarding the use of the permissive mode for authorization:

* For environments where **authorization is disabled**, this task helps you test whether it's safe to
enable the authorization.

* For environments where **authorization is enabled**, this task helps you test whether it's safe to
add a new authorization policy.

## Before you begin

To complete this task, you should first take the following actions:

* Read the [authorization concept](/docs/concepts/security/#authorization).

* Follow the instructions in the [Kubernetes quick start](/docs/setup/kubernetes/quick-start/) to
install Istio **with mutual TLS enabled**.

* Deploy the [Bookinfo](/docs/examples/bookinfo/) sample application.

> * Create service accounts and for the Bookinfo application. Run the following command to create service

Reviewer: ```suggestion
* Create service accounts for the Bookinfo application. Run the following command to create service
```

    {"level":"warn","time":"2018-08-30T21:55:53.544441Z","instance":"rbacsamplelog.logentry.istio-system","destination":"productpage","latency":"57.800056ms","permissiveResponseCode":"allowed","permissiveResponsePolicyID":"productpage-viewer","responseCode":200,"responseSize":5723,"source":"istio-ingressgateway","user":"cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account"}
    {{< /text >}}

1.  Verify that the the log shows a `responseCode` of `200` and a `permissiveResponseCode` of `allowed`
    for `productpage` service.

1.  Remove the YAML files related to enabling the permissive mode:

    {{< text bash >}}
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/productpage-policy.yaml@
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-config-on-permissive.yaml@
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
    {{< /text >}}

> 1.  Congratulations! You tested an authorization policy with permissive mode and verified it will

Reviewer: ```suggestion
1.  Congratulations! You tested an authorization policy with permissive mode and verified it
```

>     work as expected. To enable the authorization policy, follow the steps described in the

Reviewer: ```suggestion
    works as expected. To enable the authorization policy, follow the steps described in the
```

    [Enabling Istio authorization task](/docs/tasks/security/authz-http#enabling-istio-authorization).

### Test adding authorization policy

The following steps show how to test a new authorization policy with permissive mode when authorization
has already been enabled.

> 1.  Allow access to the `producepage` service by following

Reviewer: ```suggestion
1.  Allow access to the `producepage` service by following the instructions in
```

---
title: Authorization permissive mode
description: Shows how to use Authorization permissive mode.
weight: 10
keywords: [security,access-control,rbac,authorization]
---

> The authorization permissive mode is an experimental feature in Istio's 1.1 release. Its interface can change in future releases.

Author: I assume you're asking to change the `Istio's 1.1 release` to `1.1 version`, is that right?

You can skip enabling the permissive mode and directly [enable Istio authorization](/docs/tasks/security/authz-http#enabling-istio-authorization)
if you do not want to try out the permissive mode feature.

This task shows how to use authorization permissive mode in below two scenarios:

* In environment without authorization, test whether it's safe to enable authorization.

* In environment already with authorization enabled, test whether it's safe to add a new authorization policy.

## Before you begin

The activities in this task assume that you:

* Understand [authorization](/docs/concepts/security/#authorization) concepts.

> * Have set up Istio on Kubernetes **with authentication enabled** by following the instructions in the

Author: Note It should be installing Istio with authentication, not Kubernetes with authentication.

  [quick start](/docs/setup/kubernetes/quick-start/), this tutorial requires mutual TLS to work. Mutual TLS
  authentication should be enabled in the [installation steps](/docs/setup/kubernetes/quick-start/#installation-steps).

* Deploy the [Bookinfo](/docs/examples/bookinfo/) sample application.

> ### Testing whether it's safe to turn on authorization globally

Author: This is about authorization and has nothing to do with authentication. I changed to `Test enabling authorization globally`

This tasks show how to use authorization permissive mode to test whether it's safe to

> turn on authorization globally.

Author: This section is about enabling authorization globally, next section is about test an specific authorization policy.

Changed to `The following steps show you how to use authorization permissive mode to test whether it's safe to
turn on authorization globally:`

        namespaces: ["default"]
      enforcement_mode: PERMISSIVE
    EOF
    {{< /text >}}

    Point your browser at the Bookinfo `productpage` (`http://$GATEWAY_URL/productpage`), you should
    see everything works fine, same as in [preparation task](#before-you-begin).

1.  Apply YAML file for the permissive mode metric collection.

    Run the following command:

    {{< text bash >}}
    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
    logentry.config.istio.io/rbacsamplelog created
    stdio.config.istio.io/rbacsamplehandler created
    rule.config.istio.io/rabcsamplestdio created
    {{< /text >}}

1.  Send traffic to the sample application.

    For the Bookinfo sample, visit `http://$GATEWAY_URL/productpage` in your web
    browser or issue the following command:

    {{< text bash >}}
    $ curl http://$GATEWAY_URL/productpage
    {{< /text >}}

    Point your browser at the Bookinfo `productpage` (`http://$GATEWAY_URL/productpage`), you should
    see everything works fine.

1.  Verify the logs stream has been created and check `permissiveResponseCode`.

    In a Kubernetes environment, search through the `istio-telemetry`

>     pods' logs as follows:

Author: Changed to `Get the log for telemetry and search for the `permissiveResponseCode` with the following command:`

    {"level":"warn","time":"2018-08-30T21:53:41.037824Z","instance":"rbacsamplelog.logentry.istio-system","destination":"reviews","latency":"1.091670916s","permissiveResponseCode":"denied","permissiveResponsePolicyID":"","responseCode":200,"responseSize":379,"source":"productpage","user":"cluster.local/ns/default/sa/bookinfo-productpage"}
    {"level":"warn","time":"2018-08-30T21:53:41.019851Z","instance":"rbacsamplelog.logentry.istio-system","destination":"productpage","latency":"1.112521495s","permissiveResponseCode":"denied","permissiveResponsePolicyID":"","responseCode":200,"responseSize":5723,"source":"istio-ingressgateway","user":"cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account"}
    {{< /text >}}

    In the above telemetry logs,  the `responseCode` is 200 which is what user see now.
    The `permissiveResponseCode` is `denied` which is what user will see after switching
    global authorization configuration from `PERMISSIVE` mode to `ENFORCED` mode, which
    indicates the global authorization configuration will work as expected after rolling

>     to production.

Reviewer: This text fails to convey to readers what actions they should be taking. To the readers, the log is in front of them and they are seeing the log at that moment, thus the use of the future tense is confusing. The paragraph includes information that is not pertinent to the action the reader needs to take. Remove the unrelated information to make the text easier to follow. You do mention how the log changes later on, there is no need to obfuscate the action with that information at this point.
```suggestion
1. Verify that the the log shows a `responseCode` of `200` and a `permissiveResponseCode` of `denied`.
```

1.  Before rolling out a new authorization policy in production, apply it in permissive mode.
    `Note`, when global authorization configuration is in permissive mode, all policies will be in

>     permissive mode by default.

Reviewer: Is the reader running the commands in this doc applying the policy before rolling out into production? The answer is clearly yes, thus the text creates more questions than it answers.  Readers might wonder whether or not they already did something wrong. If you consider it warranted, add it as a proper warning. The text explaining what the permissive mode is and how it works should not be part of the steps. See my previous comment on that regard. Add instead the information regarding what specific authorization policy you are talking about.
```suggestion
1. Apply the `productpage-policy.yaml` authorization policy in permissive mode with the following command:
```

>     Run the following command:

Reviewer: ```suggestion
```

    {{< text bash >}}
    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/productpage-policy.yaml@
    {{< /text >}}

> 1.  Send traffic to the sample application again.

Reviewer: Just copy the previous instance of this step. Avoid using "again" as it gives readers the impression that they are not making any progress.

    For the Bookinfo sample, visit `http://$GATEWAY_URL/productpage` in your web
    browser or issue the following command:

    {{< text bash >}}
    $ curl http://$GATEWAY_URL/productpage
    {{< /text >}}

    Point your browser at the Bookinfo `productpage` (`http://$GATEWAY_URL/productpage`), you should
    see everything works fine.

> 1.  Verify the logs and check `permissiveResponseCode` again.

Reviewer: See my previous comments on how to handle steps involving the logs.

    {"level":"warn","time":"2018-08-30T21:55:53.544441Z","instance":"rbacsamplelog.logentry.istio-system","destination":"productpage","latency":"57.800056ms","permissiveResponseCode":"allowed","permissiveResponsePolicyID":"productpage-viewer","responseCode":200,"responseSize":5723,"source":"istio-ingressgateway","user":"cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account"}
    {{< /text >}}

    In telemetry logs above,  the `responseCode` is 200 which is what user see now.
    The `permissiveResponseCode` is `allowed` for productpage service, 403 for ratings
    and reviews services, which are what user will see after switching
    policy mode from `PERMISSIVE` mode to `ENFORCED` mode; the result aligns with
    [Emabling authorization for HTTP services step 1](/docs/tasks/security/authz-http#step-1-allowing-access-to-the-productpage-service).

> 1.  Remove permissive mode related yaml files:

Reviewer: The order of the words here can be confusing. Are readers supposed to remove the permissive mode or the files? There is a bit of context which is missing, why should readers remove those files?
```suggestion
1.  Remove the YAML files related to enabling the permissive mode:
```

    {{< text bash >}}
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/productpage-policy.yaml@
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-config-on-permissive.yaml@
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
    {{< /text >}}

> 1.  Now we have verified authorization will work as expected when turning it on,

Reviewer: Address the reader directly and convey a sense of accomplishment before directing them to the next task. Be explicit about what exactly the reader has accomplished and about the next steps you recommend them taking. Your text doesn't do this very effectively consider instead:
```suggestion
1.  Congratulations! You tested an authorization policy and verified it will work as expected. To enable the authorization policy, follow the steps described in the [Enabling Istio authorization task](/docs/tasks/security/authz-http#enabling-istio-authorization).
```

    it's safe following the [Enabling Istio authorization](/docs/tasks/security/authz-http#enabling-istio-authorization) to turn on authorization.

> ### Testing new authorization policy works as expected before rolling to production

Reviewer: Similarly to the previous heading, this one is too long and also fails to convey the content of the section that follows.
```suggestion
### Testing an authorization policy with authentication enabled
```

Author: Changed to `Test adding authorization policy`

This tasks shows how to use authorization permissive mode to test a new authorization policy works

> as expected in environment with authorization already enabled.

Reviewer: This text is confusing because of the grammar issues it has.
```suggestion
The following steps show how to test a new authorization policy when authorization has already been enabled.
```

> Before you start, please make sure that you have finished [Enabling authorization for HTTP services step 1](/docs/tasks/security/authz-http#step-1-allowing-access-to-the-productpage-service).

Reviewer: I would strongly encourage you to replace this text with a step and a command to verify whether authentication is enabled.
```suggestion
1. Confirm that authentication is enabled with the following command:
```

> 1.  Before applying a new policy, test it by setting its mode to permissive:

Reviewer: Please split this step as I have described in previous comments. Every step should contain a single action. This one has three: 1. Run the command, 2. Review the YAML, and 3. Verify there is an error when going to the product page.

    Run the following command:

    {{< text bash >}}
    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/details-reviews-policy-permissive.yaml@
    {{< /text >}}

    The policy is the same as defined in [allowing access to the details and
    reviews services](/docs/tasks/security/authz-http#step-2-allowing-access-to-the-details-and-reviews-services), except `PERMISSIVE` mode is set in ServiceRoleBinding.

    {{< text yaml >}}
    apiVersion: "rbac.istio.io/v1alpha1"
    kind: ServiceRoleBinding
    metadata:
      name: bind-details-reviews
      namespace: default
    spec:
      subjects:
      - user: "cluster.local/ns/default/sa/bookinfo-productpage"
      roleRef:
        kind: ServiceRole
        name: "details-reviews-viewer"
      mode: PERMISSIVE
    {{< /text >}}

    Point your browser at the Bookinfo `productpage` (`http://$GATEWAY_URL/productpage`), you should still
    see there are errors `Error fetching product details` and `Error fetching
    product reviews` on the page. These errors are expected because the policy is
    in `PERMISSIVE` mode.

> 1.  Apply YAML file for the permissive mode metric collection.

Reviewer: Be specific about what file you mean, there are many YAML files in Istio.
```suggestion
1.  Apply the `rbac-permissive-telemetry.yaml` YAML file to enable the permissive mode metric collection.
```

>     Run the following command:

Reviewer: ```suggestion
```

    {{< text bash >}}
    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
    {{< /text >}}

> 1.  Send traffic to the sample application.

Reviewer: Use the same step as in the previous section.

    For the Bookinfo sample, visit `http://$GATEWAY_URL/productpage` in your web
    browser or issue the following command:

    {{< text bash >}}
    $ curl http://$GATEWAY_URL/productpage
    {{< /text >}}

> 1.  Verify the logs and check `permissiveResponseCode` again.

Reviewer: See my previous comments regarding the steps needed for logs.

    {"level":"warn","time":"2018-08-30T22:59:42.707093Z","instance":"rbacsamplelog.logentry.istio-system","destination":"details","latency":"423.381µs","permissiveResponseCode":"allowed","permissiveResponsePolicyID":"details-reviews-viewer","responseCode":403,"responseSize":19,"source":"productpage","user":"cluster.local/ns/default/sa/bookinfo-productpage"}
    {"level":"warn","time":"2018-08-30T22:59:42.763423Z","instance":"rbacsamplelog.logentry.istio-system","destination":"reviews","latency":"237.333µs","permissiveResponseCode":"allowed","permissiveResponsePolicyID":"details-reviews-viewer","responseCode":403,"responseSize":19,"source":"productpage","user":"cluster.local/ns/default/sa/bookinfo-productpage"}
    {{< /text >}}

    In telemetry logs above, the `responseCode` is 403 for ratings
    and reviews services, which is what users see now.
    The `permissiveResponseCode` is `allowed` for ratings and reviews services,
    which is what users will see after switching policy mode from `PERMISSIVE` mode
    to `ENFORCED` mode; it indicates the new authorization policy will work as expected
    after rolling to production.

> 1.  Remove permissive mode related yaml files:

Reviewer: See my previous comment on the word order.

    {{< text bash >}}
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/details-reviews-policy-permissive.yaml@
    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
    {{< /text >}}

> 1.  Now we have verified the new policy will work as expected, it's safe

Reviewer: Please use the same text I suggested at the end of the previous section.
